tools/calyx-writer/lib.rs

```diff
@@ -0,0 +1,1077 @@
+            name: name.to_string(),
+            width: width as isize,
+        }
+    }
+
+    /// Constructs a new port with the given `name` in a component.
+    ///
+    /// Requires: `width > 0`.
+    pub fn new_in_comp<S: ToString>(name: S, width: usize) -> Self {
+        Self::new::<String, S>(None, name, width)
+    }
+
+    /// Constructs a new port with the given `parent` and `name` and with
+    /// inferred width.
+    pub fn inferred<S: ToString, T: ToString>(
+        parent: Option<S>,
+        name: T,
+    ) -> Self {
+        Self {
+            attributes: vec![],
+            parent: parent.map(|s| s.to_string()),
+            name: name.to_string(),
+            width: -1,
+        }
+    }
+
+    pub fn has_inferred_width(&self) -> bool {
+        self.width == -1
+    }
+}
+
+impl AttributeProvider for Port {
+    fn add_attribute(&mut self, attr: Attribute) {
+        self.attributes.push(attr);
+    }
+}
+
+impl Display for Port {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if let Some(parent) = &self.parent {
+            write!(f, "{}.", parent)?;
+        }
+        write!(f, "{}", self.name)
+    }
+}
+
+impl CalyxWriter for Port {
+    /// Behaves identically to [`Port::fmt`]. Please read [`CalyxWriter::write`]
+    /// for documentation on this function.
+    fn write(&self, f: &mut IndentFormatter<'_, '_>) -> fmt::Result {
+        write!(f, "{}", self)
+    }
+}
+
+/// Abstracts port functionality over components and cells.
+pub trait PortProvider {
+    /// Retrieves the port on this provider named `port`.
+    ///
+    /// Requires: `port` exists on this provider.
+    fn get(&self, port: String) -> Port;
+}
+
+/// See `calyx_ir::Cell`.
+pub struct Cell {
+    is_ref: bool,
+    attributes: Vec<Attribute>,
+    name: String,
+    /// The name of the component or primitive instantiated, e.g., "std_reg"
+    /// for the register primitive.
+    inst: String,
+    args: Vec<u64>,
+}
+
+impl Cell {
+    /// Whether this cell is a ref cell.
+    pub fn is_ref(&self) -> bool {
+        self.is_ref
+    }
+}
+
+impl PortProvider for Cell {
+    fn get(&self, port: String) -> Port {
+        Port::inferred(Some(self.name.clone()), port)
+    }
+}
+
+impl AttributeProvider for Cell {
+    fn add_attribute(&mut self, attr: Attribute) {
+        self.attributes.push(attr);
+    }
+}
+
+impl CalyxWriter for Cell {
+    fn write(&self, f: &mut IndentFormatter<'_, '_>) -> fmt::Result {
+        self.attributes.write(f)?;
+        write!(
+            f,
+            "{}{} = {}({});",
+            if self.is_ref { "ref " } else { "" },
+            self.name,
+            self.inst,
+            self.args
+                .iter()
+                .map(u64::to_string)
+                .collect::<Vec<_>>()
+                .join(", ")
+        )
+    }
+}
+
+/// A guard for an [`Assignment`].
+#[derive(PartialEq, Eq, Debug)]
+pub enum Guard {
+    None,
+    Constant(u64),
```

Comment:
I assume this is supposed to be a comparison with a constant value? I'd be a
little careful here since a constant by itself is not a wellformed guard